File: internal/domain/order.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

type OrderStatus string

const (
	OrderStatusPending   OrderStatus = "pending"
	OrderStatusPaid      OrderStatus = "paid"
	OrderStatusReady     OrderStatus = "ready"
	OrderStatusCompleted OrderStatus = "completed"
	OrderStatusCancelled OrderStatus = "cancelled"
)

type Order struct {
	ID              uuid.UUID   `json:"id"                         db:"id"`
	CustomerID      uuid.UUID   `json:"customer_id"                db:"customer_id"`
	BusinessID      uuid.UUID   `json:"business_id"                db:"business_id"`
	Items           []OrderItem `json:"items"`
	TotalAmount     float64     `json:"total_amount"               db:"total_amount"`
	Status          OrderStatus `json:"status"                     db:"status"`
	PIN             string      `json:"-"                          db:"pin"`
	StripePaymentID string      `json:"stripe_payment_id,omitempty" db:"stripe_payment_id"`
	CreatedAt       time.Time   `json:"created_at"                 db:"created_at"`
	UpdatedAt       time.Time   `json:"updated_at"                 db:"updated_at"`
}

type OrderItem struct {
	ID          uuid.UUID `json:"id"           db:"id"`
	OrderID     uuid.UUID `json:"order_id"     db:"order_id"`
	ProductName string    `json:"product_name" db:"product_name"`
	Quantity    int       `json:"quantity"     db:"quantity"`
	UnitPrice   float64   `json:"unit_price"   db:"unit_price"`
}

// OrderResponse is returned to the customer after payment — includes the PIN once.
type OrderResponse struct {
	Order
	PIN string `json:"pin,omitempty"`
}

type CreateOrderRequest struct {
	BusinessID uuid.UUID            `json:"business_id" validate:"required"`
	Items      []CreateOrderItemReq `json:"items"       validate:"required,min=1,dive"`
}

type CreateOrderItemReq struct {
	ProductName string  `json:"product_name" validate:"required"`
	Quantity    int     `json:"quantity"     validate:"required,min=1"`
	UnitPrice   float64 `json:"unit_price"   validate:"required,gt=0"`
}

type ValidatePINRequest struct {
	PIN string `json:"pin" validate:"required,len=6"`
}

File: mobile/lib/models/order.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'order.freezed.dart';
part 'order.g.dart';

enum OrderStatus {
  @JsonValue('pending') pending,
  @JsonValue('paid') paid,
  @JsonValue('ready') ready,
  @JsonValue('completed') completed,
  @JsonValue('cancelled') cancelled,
}

@freezed
class OrderItem with _$OrderItem {
  const factory OrderItem({
    required String id,
    @JsonKey(name: 'order_id') required String orderId,
    @JsonKey(name: 'product_name') required String productName,
    required int quantity,
    @JsonKey(name: 'unit_price') required double unitPrice,
  }) = _OrderItem;

  factory OrderItem.fromJson(Map<String, dynamic> json) =>
      _$OrderItemFromJson(json);
}

@freezed
class Order with _$Order {
  const factory Order({
    required String id,
    @JsonKey(name: 'customer_id') required String customerId,
    @JsonKey(name: 'business_id') required String businessId,
    required List<OrderItem> items,
    @JsonKey(name: 'total_amount') required double totalAmount,
    required OrderStatus status,
    // Only present immediately after successful payment
    String? pin,
    @JsonKey(name: 'stripe_payment_id') String? stripePaymentId,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _Order;

  factory Order.fromJson(Map<String, dynamic> json) => _$OrderFromJson(json);
}

// ─── Request models ──────────────────────────────────────────────────────────

@freezed
class CreateOrderItemRequest with _$CreateOrderItemRequest {
  const factory CreateOrderItemRequest({
    @JsonKey(name: 'product_name') required String productName,
    required int quantity,
    @JsonKey(name: 'unit_price') required double unitPrice,
  }) = _CreateOrderItemRequest;

  factory CreateOrderItemRequest.fromJson(Map<String, dynamic> json) =>
      _$CreateOrderItemRequestFromJson(json);
}

@freezed
class CreateOrderRequest with _$CreateOrderRequest {
  const factory CreateOrderRequest({
    @JsonKey(name: 'business_id') required String businessId,
    required List<CreateOrderItemRequest> items,
  }) = _CreateOrderRequest;

  factory CreateOrderRequest.fromJson(Map<String, dynamic> json) =>
      _$CreateOrderRequestFromJson(json);
}
